File: dotnet/Allmark/Render/ConsoleCodeBlockRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleCodeBlockRenderer
{
    public static Renderer Create(string style)
    {
        return new Renderer
        {
            Name = "code_block",
            Render = (node, state, first, last, decode) => Render(node, state, style),
        };
    }

    public static void Render(MarkdownNode node, RendererState state, string style)
    {
        if (state.Output.Length > 0 && state.Output[^1] != '\n')
        {
            state.Output.Append('\n');
        }
        state.Output.Append($"{style}┌─{RenderToConsole.AnsiReset}\n");
        foreach (var line in node.Content.Split('\n'))
        {
            state.Output.Append($"{style}│{RenderToConsole.AnsiReset} {line}\n");
        }
        state.Output.Append($"{style}└─{RenderToConsole.AnsiReset}\n");
    }
}

File: swift/Sources/allmark/RenderConsole/ConsoleCodeBlockRenderer.swift
import Foundation

@MainActor
let consoleCodeBlockRenderer = Renderer(
	name: "code_block",
	render: renderConsoleCodeBlock
)

@MainActor
func renderConsoleCodeBlock(_ node: MarkdownNode, _ state: inout RendererState, _ first: Bool?, _ last: Bool?, _ decode: Bool?) {
	let styles = getConsoleStyles()
	let style = styles["codeBlock"] ?? ""
	if !state.output.isEmpty && !state.output.hasSuffix("\n") {
		state.output += "\n"
	}
	state.output += "\(style)┌─\(ansiReset)\n"
	for line in node.content.split(separator: "\n", omittingEmptySubsequences: false) {
		state.output += "\(style)│\(ansiReset) \(line)\n"
	}
	state.output += "\(style)└─\(ansiReset)\n"
}
